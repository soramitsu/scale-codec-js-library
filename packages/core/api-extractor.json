{
  "extends": "../../api-extractor.json",
  "mainEntryPointFilePath": "<projectFolder>/types/core/src/lib.d.ts",
  "dtsRollup": {
    "untrimmedFilePath": "dist/lib.d.ts"
  }
}
