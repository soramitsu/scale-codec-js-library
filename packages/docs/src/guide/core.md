# Core

This library is a pure low-level performant implementation of SCALE.

## Supported types

-   **Primitive:**

    | Spec    | JS Type             | Details                                                                                                                                                             |
    | ------- | ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
    | Int     | `number`, `bigint`  | signed & unsigned, 8/16/32/64/128/etc bits. For ints with bits less than or equal 32 both `number` & `bigint` could be used, for 64+ bits integers - only `bigint`. |
    | Compact | `number`, `bigint`  | -                                                                                                                                                                   |
    | String  | `string`            | -                                                                                                                                                                   |
    | Bool    | `boolean`           | -                                                                                                                                                                   |
    | Void    | `null`, `undefined` | `()` in Rust. Unfortunately, JavaScript doesn't have zero-cost abstractions, so this codec could be used to handle it.                                              |

-   **Higher-order:**

    | Spec   | JS Type                                        |
    | ------ | ---------------------------------------------- |
    | Array  | `Array`                                        |
    | Vector | `Array`                                        |
    | Tuple  | `Array`, but tuple in TypeScript               |
    | Set    | `Set`                                          |
    | Map    | `Map`                                          |
    | Struct | Plain Object                                   |
    | Enum   | `Enum` from [`@scale-codec/enum`](/guide/enum) |

-   **Special:**

    -   Efficient codec for arrays of bytes, or `[u8; x]` in Rust and `Uint8Array` in JS
    -   Efficient codec for vectors of bytes, or `Vec<u8>` in Rust and `Uint8Array` in JS
    -   `OptionBool`

## Package

Available on NPM:

```bash
npm i @scale-codec/core
```

## Core Concept

All codec functions in the library are built on top of these core types:

```ts
interface Walker {
    u8: Uint8Array
    idx: number
    view: DataView
}

interface Encode<T> {
    (value: T, walker: Walker): void
    sizeHint: (value: T) => number
}

type Decode<T> = (walker: Walker) => T
```

While **encoding**, firstly you should compute the value size prediction with a `.sizeHint(value)` function of `Encode` function, allocate buffer for it, put it into Walker interface and the pass it to the Encode function itself, which will mutate it's buffer and offset index. For example, here is a boolean encode function:

```ts
const encodeBool: Encode<boolean> = (value, walker) => {
    walker.u8[walker.idx++] = value ? 1 : 0
}

encodeBool.sizeHint = () => 1
```

While **decoding**, each decode function reading from Walker's buffer and incrementing it's offset index. For example, here is a boolean decode function:

```ts
const decodeBool: Decode<boolean> = (walker) => {
    return walker.u8[walker.idx++] === 1
}
```

Core library provides a very useful `WalkerImpl` utility to work with codecs:

```ts
const encoded: Uint8Array = WalkerImpl.encode(false, encodeBool)

const decoded: boolean = WalkerImpl.decode(encoded, decodeBool)
```

## Typing Tips

### Tuples

You can use TypeScript tuples to define tuple codecs safely:

```ts
import {
    createTupleEncoder,
    encodeU8,
    encodeStr,
} from '@scale-codec/core'

// (u8, str)
type MyTuple = [number, string]

const encoder = createTupleEncoder<MyTuple>([encodeU8, encodeStr])
```

### Structs

Struct codecs type definition is not completely automatic due to performance reasons and TypeScript limitations. Let's define a sample structure.

Definition in Rust:

```rust
use parity_scale_codec::{Encode, Decode};

#[derive(Encode, Decode)]
struct Message {
    author: String,
    timestamp: u128
}
```

Codecs with core library:

```ts
import {
    createStructEncoder,
    createStructDecoder,
    encodeU128,
    decodeU128,
    encodeStr,
    decodeStr,
} from '@scale-codec/core'

interface Message {
    author: string
    timestamp: bigint
}

const encoder = createStructEncoder<Message>([
    ['author', encodeStr],
    ['timestamp', encodeU128],
])

const decoder = createStructDecoder<Message>([
    ['author', decodeStr],
    ['timestamp', decodeU128],
])
```

Tuples passed to factories are typed only partially, i.e. TypeScript does not provide a guarantee that passed encoders/decoders schema will result into correct codecs. You should ensure by yourself that:

-   The order of fields is correct
-   All fields are defined

### Enums

::: tip
Enums handling is based on `@scale-codec/enum` package. More info at [Enums guide](./enum.md).
:::

::: tip
Core package re-exports everything from `@scale-codec/enum`, so:

```ts
import { Enum, Option /* ... */ } from '@scale-codec/core'
```

:::

Enums codecs _creation_ is even less strongly typed as struct ones. You are responsible to provide the correct schema. Let's define a sample enum.

In Rust:

```rust
enum Event {
    Focus,
    KeyPress(String)
}
```

In TypeScript:

```ts
import { Enum } from '@scale-codec/core'

type Event = Enum<'Focus' | ['KeyPress', string]>
```

Encoder:

```ts
import { createEnumEncoder, encodeStr } from '@scale-codec/core'

const encoder = createEnumEncoder<Event>({
    // map from <tag name> to <discriminant> for empty variants
    Focus: 0,
    // or [<discriminant>, <encode>] for valuable variants
    KeyPress: [1, encodeStr],
})
```

Decoder:

```ts
import { createEnumDecoder, decodeStr } from '@scale-codec/core'

const decoder = createEnumDecoder<Event>({
    // Map from <discriminant> to <tag name> for empty variants
    0: 'Focus',
    // or [<tag name>, <decode>] for valuable variants
    1: ['KeyPress', decodeStr],
})
```

Such approach is not very convenient to use, but it is worth it due to performance reasons.

Enums handling is more low-level then other tools because of typing issues. Thus, the responsibility to define encode & decode schema correctly is on the end user.

## Wrapping Up

This page doesn't describe all functions that it includes. For the more detailed API please refer to the related section.

Anyway, if you need to build a large namespace and avoid a lot of symmetric "encode/decode" functions, use tracking functionality for debugging with almost the same performance, you probably need to explore the next section.
