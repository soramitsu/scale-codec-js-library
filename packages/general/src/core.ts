export type CodecType<V, E = {}> = {
    encode: (value: V) => Uint8Array;
    decode: (buffer: Uint8Array) => V;
    // create?: (something: Default) => V;
    // createFromVoid?: () => V;
} & E;

export type CodecTypeValue<C extends CodecType<unknown>> = C extends CodecType<infer V> ? V : never;

// export type NamescapeDefault = Record<string, CodecType<unknown>>;

export type NamespaceOptions<N extends {}> = {
    [K in keyof N]: CodecTypeOptions<N, N[K]>;
};

export interface Root<N extends {}> {
    lookup: RootLoolupFn<N>;
    encode: RootEncodeFn<N>;
    decode: RootDecodeFn<N>;
}

export type RootLoolupFn<N extends {}> = <K extends keyof N>(type: K) => CodecType<N[K]>;

export type RootEncodeFn<N extends {}> = <K extends keyof N>(type: K, value: N[K]) => Uint8Array;

export type RootDecodeFn<N extends {}> = <K extends keyof N>(type: K, buffer: Uint8Array) => N[K];

// definition internal, frame worked
export type CodecTypeOptions<N extends {}, V extends any, R = Root<N>> = {
    encode: (root: R, value: V) => Uint8Array;
    decode: (root: R, buffer: Uint8Array) => V;
    // create?: <V>(root: R, value: V) => T;
    // createFromVoid?: (root: R) => T;
};

function compileTypeDefinition<N extends {}, V>(root: Root<N>, typeOptions: CodecTypeOptions<N, V>): CodecType<V> {
    const { encode, decode } = typeOptions;
    return {
        encode: (v) => encode(root, v),
        decode: (b) => decode(root, b),
    };
}

type NamespaceToCodecTypes<N extends {}> = {
    [K in keyof N]: CodecType<N[K]>;
};

export function createRoot<N extends {}>(namescape: NamespaceOptions<N>): Root<N> {
    const root: Root<N> = {} as any;

    const types: NamespaceToCodecTypes<N> = Object.fromEntries(
        (Object.entries(namescape) as [keyof N, CodecTypeOptions<N, any>][]).map(([typeName, options]) => [
            typeName,
            compileTypeDefinition(root, options),
        ]),
    ) as any;

    root.lookup = (type) => types[type];
    root.encode = (type, val) => root.lookup(type).encode(val);
    root.decode = (type, buff) => root.lookup(type).decode(buff);

    return root;
}

// class RootImpl<N extends NamescapeDefault> implements Root<N> {
//     private types: N;

//     constructor(defs: NamespaceOptions<N>) {
//         this.types = Object.fromEntries(
//             (Object.entries(defs) as [keyof N, TypeOptions<N, any>][]).map(([typeName, options]) => [
//                 typeName,
//                 compileTypeDefinition(this, options),
//             ]),
//         ) as any;
//     }

//     lookup<K extends keyof N>(type: K): N[K] {
//         return this.types[type];
//     }

//     encode<K extends keyof N>(type: K, value: CodecTypeValue<N[K]>): Uint8Array {
//         return this.lookup(type).encode(value);
//     }

//     decode<K extends keyof N>(type: K, buffer: Uint8Array): CodecTypeValue<N[K]> {
//         return this.lookup(type).decode(buffer);
//     }
// }

{
    type MyTypes = {
        Id: {
            name: string;
            domain: string;
        };
        Account: {
            id: MyTypes['Id'];
        };
    };

    // interface Id {
    //     name: string;
    //     domain: string;
    // }

    // interface Account {
    //     id: Id;
    // }

    // type Namespace = {
    //     [K in keyof MyTypes]: CodecType<MyTypes[K]>;
    // };

    const root = createRoot<MyTypes>({
        Id: {
            encode(root, { name, domain }) {
                console.log('encoding %o & %o', name, domain);
                // const name = id;
                return new Uint8Array();
            },
            decode(root, buff) {
                return { name: 'test', domain: 'puff' };
            },
        },
        Account: {
            encode: (root, { id }) => {
                return root.lookup('Id').encode(id);
                // root.lookup('Id')
                // return root.encode('Id', id);
            },
            decode: (root, buff) => {
                const id = root.lookup('Id').decode(buff);
                return { id };
            },
        },
    });

    // root.lookup('Id').
}

{
    type CodecNumType = 'signed' | 'unsigned';

    class CodecNumber {
        public constructor(value: number, bits: number, type: CodecNumType) {}

        encode(): Uint8Array {}
    }

    function createCodecNumber(bits: number, type: CodecNumType): CodecTypeOptions<any, CodecNumber> {
        return {
            decode: (root, buff) => new CodecNumber(4123, bits, type),
            encode: (root, val) => val.encode(),
        };
    }

    const root = createRoot<{
        u8: CodecNumber;
        u16: CodecNumber;
        u32: CodecNumber;
        i8: CodecNumber;
        i16: CodecNumber;
        i32: CodecNumber;
    }>({
        u8: createCodecNumber(8, 'unsigned'),
        u16: createCodecNumber(16, 'unsigned'),
        u32: createCodecNumber(32, 'unsigned'),
        i8: createCodecNumber(8, 'signed'),
        i16: createCodecNumber(16, 'signed'),
        i32: createCodecNumber(32, 'signed'),
    });

    const buff = root.lookup('u8').encode(new CodecNumber(4152, 412, 'unsigned'));
}

// custom type options
{
    type NS = {
        SmartString: CodecTypeOptions<NS, string> & { fromHex(hex: string): string };
    };

    const root = createRoot<NS>(null as any);

    root.lookup('SmartString').fromHex();

    type Base = {
        foo: string;
    };

    interface BaseExtended extends Base {
        bar: number;
    }

    type BaseExtension = Omit<BaseExtended, keyof Base>;
}

// enum
{
}
// type TypeCompiled<
//     Defs extends Definitions<any, any>,
//     K extends keyof Defs,
//     Def extends TypeDef<any, any> = Defs[K],
// > = Def extends TypeDef<any, infer T>
//     ? {
//           encode: (value: T) => Uint8Array;
//           decode: (buffer: Uint8Array) => T;
//       } & (Def extends { create: (root: any, value: infer V) => T } ? { create: (value: V) => T } : {})
//     : //  & (
//       //     Def extends { createFromVoid: (root: any) }
//       // )
//       never;

// type DefinedTypes<N> = keyof N & string;

// class Root<Defs extends Definitions<Namespace, Root<Namespace, any>>> {
//     constructor(types: Defs) {}

//     lookup<K extends DefinedTypes<Namespace>>(type: K): TypeCompiled<Defs, K> {
//         return null;
//     }
// }

// const root = new Root({
//     Id: {
//         encode: () => new Uint8Array(),
//         decode: () => 'string',
//     },
// });

// interface Codec {
//     decode(encodedBytes: Uint8Array): unknown;

// }

// export function defineDecoder<T extends ScaleEncoder>(
//     something: T & {
//         decode(bytes: Uint8Array): T;
//     },
// ): T & ScaleDecoder<T> {
//     return { ...something, [ScaleSymbolDecode]: something.decode };
// }

// export function defineEncoder<T extends ScaleEncoder>(
//     something: T & {
//         encode
//     }
// )
