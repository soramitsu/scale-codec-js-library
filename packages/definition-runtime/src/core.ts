import { Decode, Encode, Walker, WalkerImpl, encodeFactory } from '@scale-codec/core'
import { trackDecode } from './tracking'

/**
 * General interface for any codec
 */
export interface Codec<Encoded, Decoded = Encoded> {
  encodeRaw: Encode<Encoded>
  decodeRaw: Decode<Decoded>
  fromBuffer: (src: ArrayBufferView) => Decoded
  toBuffer: (value: Encoded) => Uint8Array
  // name: () => string
}

export type CodecAny = Codec<any, any>

// /**
//  * General {@link Codec} implementation that includes fragments & tracking functionality
//  */
// export class CodecImpl<E, D = E> implements Codec<E, D> {
//     public encodeRaw: Encode<E>
//     public decodeRaw: Decode<D>
//     public name: () => string
//     public fromBuffer: (src: ArrayBufferView) => D
//     public toBuffer: (value: E) => Uint8Array

//     public constructor(name: string, encode: Encode<E>, decode: Decode<D>) {
//         this.encodeRaw = encode
//         this.decodeRaw = (walker) => trackDecode(name, walker, decode)
//         this.name = () => name
//         this.toBuffer = (value) => WalkerImpl.encode(value, this.encodeRaw)
//         this.fromBuffer = (src) => WalkerImpl.decode(src, this.decodeRaw)
//     }

//     // public fromBuffer(this: this, src: ArrayBufferView): D {
//     //     return WalkerImpl.decode(src, this.decodeRaw)
//     // }

//     // public toBuffer(this: this, value: E): Uint8Array {
//     //     return WalkerImpl.encode(value, this.encodeRaw)
//     // }

//     // public name(this: this): string {
//     //     return this._name
//     // }
// }

export function trackableCodec<E, D = E>(name: string, encode: Encode<E>, decode: Decode<D>): Codec<E, D> {
  const decodeTracked: Decode<D> = (walker) => trackDecode(name, walker, decode)

  return {
    // name: () => name,
    encodeRaw: encode,
    decodeRaw: decodeTracked,
    toBuffer: (value) => WalkerImpl.encode(value, encode),
    fromBuffer: (src) => WalkerImpl.decode(src, decodeTracked),
  }
}

export type CodecValueEncodable<T extends Codec<any>> = T extends Codec<infer E, any> ? E : never

export type CodecValueDecoded<T extends Codec<any>> = T extends Codec<any, infer D> ? D : never

// /**
//  * Special {@link Codec} implementation to that wraps another codec and dispatches it via its getter.
//  * With this utility it is easy to implement cyclic dependencies between codecs.
//  *
//  * @remarks
//  *
//  * TODO optimize and cache got codec after the first dispatch?
//  */
// export class DynCodec<C extends CodecAny> implements Codec<CodecValueEncodable<C>, CodecValueDecoded<C>> {
//     public encodeRaw: Encode<CodecValueEncodable<C>>

//     public decodeRaw: Decode<CodecValueDecoded<C>>

//     public readonly codecGetter: () => C

//     public constructor(getter: () => C) {
//         this.codecGetter = getter

//         this.encodeRaw = encodeFactory(
//             (val, walker) => getter().encodeRaw(val, walker),
//             (val) => getter().encodeRaw.sizeHint(val),
//         )

//         this.decodeRaw = (walker) => getter().decodeRaw(walker)
//     }

//     public fromBuffer(this: this, src: ArrayBufferView): CodecValueDecoded<C> {
//         return this.codecGetter().fromBuffer(src)
//     }

//     public toBuffer(this: this, value: CodecValueEncodable<C>): Uint8Array {
//         return this.codecGetter().toBuffer(value)
//     }

//     public name(this: this): string {
//         return this.codecGetter().name()
//     }
// }

// /**
//  * See {@link DynCodec}. This function is a shorthand factory that is more minimize-friendly than `new DynCodec(...)`
//  */
// export function dynCodec<C extends CodecAny>(getter: () => C): DynCodec<C> {
//     return new DynCodec(getter)
// }

export function dynCodec<C extends CodecAny>(getter: () => C): Codec<CodecValueEncodable<C>, CodecValueDecoded<C>> {
  let codec: C | undefined

  const getCodec = (): C => {
    if (!codec) {
      codec = getter()
    }
    return codec
  }

  return {
    // name: () => (codec ?? getCodec()).name(),
    encodeRaw: encodeFactory(
      (val, walker) => (codec ?? getCodec()).encodeRaw(val, walker),
      (val) => (codec ?? getCodec()).encodeRaw.sizeHint(val),
    ),
    decodeRaw: (walker) => (codec ?? getCodec()).decodeRaw(walker),
    toBuffer: (value) => (codec ?? getCodec()).toBuffer(value),
    fromBuffer: (src) => (codec ?? getCodec()).fromBuffer(src),
  }
}
