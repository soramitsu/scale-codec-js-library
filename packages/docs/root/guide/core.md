# Core package

`@scale-codec/core` package contains all necessary low-level tools to perform encoding and decoding between SCALE binaries and their JS representations.

## Contents

[[toc]]

## Supported types

-   Primitives
    -   Integers - signed & unsigned, 8/16/32/64/128 bits, as [`JSBI` (`BigInt` replacement for compatibility)](https://github.com/GoogleChromeLabs/jsbi), and...
        -   Compact (special integers representation) - as `JSBI` too
    -   String - as native JS `String`
    -   Boolean - as native JS `Boolean`
    -   Void (just for consistence) - as `null`
-   Complex
    -   Fixed-length Array - as JS `Array`, and...
        -   `[u8; len]` as JS `Uint8Array`
    -   Vector - as JS `Array`, and...
        -   `Vec<u8>` as JS `Uint8Array`
    -   Set - as native JS `Set`
    -   Map - as native JS `Map`
    -   Struct - as JS object
    -   Tuple - as JS `Array`
    -   Enum (+ special codec for OptionBool) - as `Enum` from [`@scale-codec/enum` package](./enum)

## Usage examples

<script setup>
import BigIntPlayground from './components/BigIntPlayground.vue'
import BigIntEncode from './components/BigIntEncode.vue'
import CompactPlayground from './components/CompactPlayground.vue'
import CompactEncode from './components/CompactEncode.vue'
import StrEncode from './components/StrEncode.vue'
import CoreStructResult from './components/CoreStructResult.vue'
</script>

### Integers

```ts
import { encodeBigInt, JSBI } from '@scale-codec/core';

const num = JSBI.BigInt(5_012_009);
const encoded = encodeBigInt(num, {
    bits: 32,
    signed: false,
    endianness: 'le',
});

console.log(encoded);
```

Result:

<BigIntEncode :bits="32" endianness="le" num="5012009" />

[Playground](#bigint-playground)

### Compact

```ts
import { encodeCompact, JSBI } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

for (const num of [1, 34_000, 891_000_000, '24141828384918234']) {
    const encoded = encodeCompact(JSBI.BigInt(num));
    console.log('%s: %s', num, hexifyBytes(encoded));
}
```

Result:

<p><template v-for="x in [1, 34_000, 891_000_000, '24141828384918234']">{{ x }}: <code><CompactEncode :num="String(x)" hex /></code><br></template></p>

[Playground](#compact-playground)

### Strings

```ts
import { encodeStr } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

console.log(hexifyBytes(encodeStr('Лев Николаевич Толстой')));
```

Result:

<code><StrEncode val="Лев Николаевич Толстой" /></code>

### Other primitives

```ts
import { encodeBool, decodeBool, encodeVoid, decodeVoid } from '@scale-codec/core';

function assertEq(a: any, b: any) {
    // asserting deep equality between a & b...
}

assertEq(encodeBool(false), new Uint8Array([0]));
assertEq(encodeBool(true), new Uint8Array([1]));
assertEq(decodeBool(new Uint8Array([1])), [true, 1]);

assertEq(decodeVoid(new Uint8Array()), [null, 0]);
assertEq(decodeVoid(new Uint8Array([1, 2, 4, 5])), [null, 0]);
```

### Struct

Definition in Rust:

```rust
use parity_scale_codec::{Encode, Decode};

#[derive(Encode, Decode)]
struct Message {
    author: String,
    timestamp: u128
}
```

Definition & encoding in JavaScript:

```ts
import { encodeStruct, JSBI, encodeStrCompact, encodeBigInt } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

interface Message {
    author: string;
    timestamp: JSBI;
}

const msg: Message = {
    author: 'Clara',
    timestamp: JSBI.BigInt('16488182899412'),
};

const msgEncoded = encodeStruct(
    msg,
    {
        author: encodeStrCompact,
        timestamp: (v) =>
            encodeBigInt(v, {
                bits: 128,
                signed: false,
                endianness: 'le',
            }),
    },
    ['author', 'timestamp'],
);

console.log(hexifyBytes(msgEncoded));
```

Output:

<code><CoreStructResult /></code>

### Enum

::: tip
Enums handling is based on `@scale-codec/enum` package. More info at [Enums guide](./enum.md).
:::

Enums handling is more low-level then other tools because of typing issues. Thus, the responsibility to define encode & decode schema correctly is on the end user.

```ts
import { Enum, Valuable, encodeEnum, decodeEnum, encodeStr, decodeStr } from '@scale-codec/core';

type Event = Enum<{
    Focus: null;
    KeyPress: Valuable<string>;
}>;

const myEvent = decodeEnum<Event>(new Uint8Array(/* ... */), {
    // map from discriminants to variant names
    // + optional decode fn for valuable variants
    0: { v: 'Focus' },
    1: { v: 'KeyPress', decode: decodeStr },
});

const encoded = encodeEnum(myEvent, {
    // map from variant names to discriminants
    // + optional encode fn for valuable variants
    Focus: { d: 0 },
    KeyPress: { d: 1, encode: encodeStr },
});
```

### Array, Vec, Tuple

```ts
import { decodeTuple, decodeArray, decodeVec, decodeStr, decodeBool } from '@scale-codec/core';

// [String; 10]
const strings = decodeArray(new Uint8Array(/* ... */), decodeStr, 10);

// (String, bool)
const [str, bool] = decodeTuple(new Uint8Array(/* ... */), [decodeStr, decodeBool]);

// Vec<bool>
const booleans = decodeVec(new Uint8Array(/* ... */), decodeBool);
```

### Map & Set

```ts
import { encodeSet, encodeMap, encodeStr, encodeBool } from '@scale-codec/core';

encodeMap(
    new Map<string, boolean>([
        ['foo', false],
        ['bar', true],
    ]),
    encodeStr,
    encodeBool,
);

encodeSet(new Set<string>(['a', 'b', 'c']), encodeStr);
```

## Play

### BigInt Playground

<BigIntPlayground class="mt-4" />

### Compact Playground

<CompactPlayground class="mt-4" />

## Also

-   [API](../api/core)

### Why `JSBI` for numbers?

Because it can be easily migrated to the native BigInt in the future with a [special JSBI's Babel plugin](https://github.com/GoogleChromeLabs/babel-plugin-transform-jsbi-to-bigint).
