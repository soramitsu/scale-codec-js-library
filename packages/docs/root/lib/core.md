# Core package

`@scale-codec/core` package contains all necessary low-level tools to perform encoding and decoding between SCALE binaries and their JS representations.

## Contents

[[toc]]

## Supported types

-   Primitives
    -   Integers - signed & unsigned, 8/16/32/64/128 bits, as [`JSBI` (`BigInt` replacement for compatibility)](https://github.com/GoogleChromeLabs/jsbi), and...
        -   Compact (special integers representation) - as `JSBI` too
    -   String - as native JS `String`
    -   Boolean - as native JS `Boolean`
    -   Void (just for consistence) - as `null`
-   Complex
    -   Fixed-length Array - as JS `Array`, and...
        -   `[u8; len]` as JS `Uint8Array`
    -   Vector - as JS `Array`, and...
        -   `Vec<u8>` as JS `Uint8Array`
    -   Set - as native JS `Set`
    -   Map - as native JS `Map`
    -   Struct - as JS object
    -   Tuple - as JS `Array`
    -   Enum (+ special codec for OptionBool) - as `Enum` from [`@scale-codec/enum` package](./enum)

## Usage examples

<script setup>
import BigIntPlayground from './components/BigIntPlayground.vue'
import BigIntEncode from './components/BigIntEncode.vue'
import CompactPlayground from './components/CompactPlayground.vue'
import CompactEncode from './components/CompactEncode.vue'
import StrEncode from './components/StrEncode.vue'
import CoreStructResult from './components/CoreStructResult.vue'
</script>

### Integers

```ts
import { encodeBigInt, JSBI } from '@scale-codec/core';

const num = JSBI.BigInt(5_012_009);
const encoded = encodeBigInt(num, {
    bits: 32,
    signed: false,
    endianness: 'le',
});

console.log(encoded);
```

Result:

<BigIntEncode :bits="32" endianness="le" num="5012009" />

[Playground](#bigint-playground)

### Compact

```ts
import { encodeCompact, JSBI } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

for (const num of [1, 34_000, 891_000_000, '24141828384918234']) {
    const encoded = encodeCompact(JSBI.BigInt(num));
    console.log('%s: %s', num, hexifyBytes(encoded));
}
```

Result:

<div><template v-for="x in [1, 34_000, 891_000_000, '24141828384918234']">{{ x }}: <CompactEncode :num="String(x)" hex /><br></template></div>

[Playground](#compact-playground)

### Strings

```ts
import { encodeStrCompact } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

console.log(hexifyBytes(encodeStrCompact('Лев Николаевич Толстой')));
```

Result:

<StrEncode val="Лев Николаевич Толстой" />

### Other primitives

```ts
import { encodeBool, decodeBool, encodeVoid, decodeVoid } from '@scale-codec/core';

function assertEq(a: any, b: any) {
    // asserting deep equality between a & b...
}

assertEq(encodeBool(false), new Uint8Array([0]));
assertEq(encodeBool(true), new Uint8Array([1]));
assertEq(decodeBool(new Uint8Array([1])), [true, 1]);

assertEq(decodeVoid(new Uint8Array()), [null, 0]);
assertEq(decodeVoid(new Uint8Array([1, 2, 4, 5])), [null, 0]);
```

### Struct

Definition in Rust:

```rust
use parity_scale_codec::{Encode, Decode};

#[derive(Encode, Decode)]
struct Message {
    author: String,
    timestamp: u128
}
```

Definition & encoding in JavaScript:

```ts
import { encodeStruct, JSBI, encodeStrCompact, encodeBigInt } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

interface Message {
    author: string;
    timestamp: JSBI;
}

const msg: Message = {
    author: 'Clara',
    timestamp: JSBI.BigInt('16488182899412'),
};

const msgEncoded = encodeStruct(
    msg,
    {
        author: encodeStrCompact,
        timestamp: (v) =>
            encodeBigInt(v, {
                bits: 128,
                signed: false,
                endianness: 'le',
            }),
    },
    ['author', 'timestamp'],
);

console.log(hexifyBytes(msgEncoded));
```

Output:

<CoreStructResult />

### More example

Todo for:

-   Map
-   Set
-   Tuple
-   Array & Vec
-   Enum

## Play

### BigInt Playground

<BigIntPlayground class="mt-4" />

### Compact Playground

<CompactPlayground class="mt-4" />

## Also

-   [API](../api/core)
