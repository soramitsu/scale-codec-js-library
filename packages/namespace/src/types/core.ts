import { Codec } from '@scale-codec/core'

export type NamespaceDefinitions<N> = {
    [K in keyof N]: NamespaceCodec<N[K], N>
}

export type NamespaceCodec<T, N> = Codec<T> | ContextSensitiveCodec<T, N>

export interface Namespace<N> {
    encode: <K extends keyof N>(ref: K, value: N[K]) => Uint8Array
    decode: <K extends keyof N>(ref: K, bytes: Uint8Array) => N[K]
}

export type ContextSensitiveCodec<T, N> = (ctx: CodecSetupContext<N>) => Codec<T>

export interface CodecSetupContext<N> {
    dynCodec: <K extends keyof N>(ref: K) => Codec<N[K]>
}

export type CompatibleNamespaceKeys<N, T> = {
    [K in keyof N]: T extends N[K] ? K : never
}[keyof N]
