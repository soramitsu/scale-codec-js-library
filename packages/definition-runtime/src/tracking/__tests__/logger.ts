import { WalkerImpl } from '@scale-codec/core'
import { setCurrentTracker, Logger, trackDecode } from '../index'

function runSuccessTrack() {
    WalkerImpl.decode<string>(new Uint8Array([0, 1, 2, 3]), (walker) => {
        return trackDecode('First', walker, (walker) => {
            return trackDecode('Second', walker, (walker) => {
                walker.idx += 4
                return 'Result'
            })
        })
    })
}

function runFailureTrack() {
    expect(() =>
        WalkerImpl.decode(new Uint8Array([4, 2, 3, 1]), (walker) =>
            trackDecode('Whoosh', walker, (walker) =>
                // eslint-disable-next-line max-nested-callbacks
                trackDecode('Shoowh', walker, (walker) => {
                    throw new Error('Expected inner error')
                }),
            ),
        ),
    ).toThrowError('Expected inner error')
}
const noop = () => {}
let consoleErrorMock: jest.SpyInstance
let consoleDebugMock: jest.SpyInstance

beforeEach(() => {
    consoleErrorMock = jest.spyOn(console, 'error').mockImplementation(noop)
    consoleDebugMock = jest.spyOn(console, 'debug').mockImplementation(noop)
    setCurrentTracker(null)
})

afterEach(() => {
    jest.resetAllMocks()
    setCurrentTracker(null)
})

afterAll(() => {
    jest.restoreAllMocks()
})

test('Prints debug log if related prop is set', () => {
    new Logger({
        logDecodeSuccesses: true,
    }).mount()

    runSuccessTrack()

    expect(consoleErrorMock).not.toBeCalled()
    expect(consoleDebugMock).toBeCalledTimes(1)
    expect(consoleDebugMock.mock.calls[0]).toMatchSnapshot()
})

test("Doesn't print debug logs by default", () => {
    new Logger().mount()

    runSuccessTrack()

    expect(consoleErrorMock).not.toBeCalled()
    expect(consoleDebugMock).not.toBeCalled()
})

test('Prints error log by default', () => {
    new Logger().mount()

    runFailureTrack()

    expect(consoleDebugMock).not.toBeCalled()
    expect(consoleErrorMock).toBeCalledTimes(1)
    expect(consoleErrorMock.mock.calls[0]).toMatchSnapshot()
})

test("Doesn't print error if prop is set", () => {
    new Logger({ logDecodeErrors: false }).mount()

    try {
        runFailureTrack()
    } catch {}

    expect(consoleDebugMock).not.toBeCalled()
    expect(consoleErrorMock).not.toBeCalled()
})

// test('Changes bytesLimit if provided some', () => {
//     new Logger({ bytesPrintLimit: 2, logDecodeSuccesses: true }).mount()

//     runSuccessTrack()
//     runFailureTrack()

//     expect(consoleDebugMock.mock.calls[0]).toMatchSnapshot()
//     expect(consoleErrorMock.mock.calls[0]).toMatchSnapshot()
// })
