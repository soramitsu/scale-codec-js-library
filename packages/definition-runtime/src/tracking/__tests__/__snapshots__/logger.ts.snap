// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prints debug log if related prop is set 1`] = `
Array [
  "[SCALE] Decode of \\"First\\" succeed

Decode steps:

First
    Walk: <offset: 0..4 (+4); 00 01 02 03>
    Result: %O
    Child steps: 1
First / Second
    Walk: <offset: 0..4 (+4); 00 01 02 03>
    Result: %O
    Child steps: 0
",
  "Result",
  "Result",
]
`;

exports[`Prints error log by default 1`] = `
Array [
  "[SCALE] Decode of \\"Whoosh\\" failed with error: Error: Expected inner error

Decode steps:

Whoosh
    Walk: <offset: 0; 04 02 03 01>
    Result: <not computed>
    Child steps: 1
Whoosh / Shoowh
    Walk: <offset: 0; 04 02 03 01>
    Result: ERROR - %s
    Child steps: 0
",
  [Error: Expected inner error],
]
`;
